packages/typegpu/src/core/function/tgpuVertexFn.ts

```diff
@@ -119,14 +126,29 @@ function createVertexFn(
 
     $name(newLabel: string): This {
       core.label = newLabel;
-      if (isWgslStruct(outputType)) {
+      if (isNamable(outputType)) {
         outputType.$name(`${newLabel}_Output`);
       }
       return this;
     },
 
     '~resolve'(ctx: ResolutionCtx): string {
-      return core.resolve(ctx, '@vertex ');
+      if (typeof implementation === 'string') {
+        return core.resolve(ctx, '@vertex ');
+      }
+
+      const generationCtx = ctx as GenerationCtx;
+      if (generationCtx.callStack === undefined) {
+        throw new Error(
+          'Cannot resolve a TGSL function outside of a generation context',
+        );
+      }
+
+      generationCtx.callStack.push(outputType);
+      const resolved = core.resolve(ctx, '@vertex ');
+      generationCtx.callStack.pop();
+
+      return resolved;
```

Comment:
I don't know which one i prefer 🤔
```suggestion
      try {
        generationCtx.callStack.push(outputType);
        return core.resolve(ctx, '@vertex ');
      } finally {
        generationCtx.callStack.pop();
      }
```

Reply:
i like the one with finally, but I don't insist

Reply:
I think I prefer the one with finally as well 👌